src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -710,58 +710,58 @@ private void pruneDeletedTombstones() {
-                 * The way we implement upgrades is a bit hackish in the sense that we set an instance
-                 * variable and that this setting will thus apply to the next forced merge that will be run.
-                 * This is ok because (1) this is the only place we call forceMerge, (2) we have a single
-                 * thread for optimize, and the 'optimizeMutex' guarding this code, and (3) ConcurrentMergeScheduler
-                 * syncs calls to findForcedMerges.
-                 */
-                MergePolicy mp = indexWriter.getConfig().getMergePolicy();
-                assert mp instanceof ElasticsearchMergePolicy : "MergePolicy is " + mp.getClass().getName();
+        ensureOpen();
+        if (optimizeLock.tryLock()) { // we do NOT acquire the readlock here since we are waiting on the merges to finish - that's fine since the IW.rollback should interrupt all the
+            /*
+             * The way we implement upgrades is a bit hackish in the sense that we set an instance
+             * variable and that this setting will thus apply to the next forced merge that will be run.
+             * This is ok because (1) this is the only place we call forceMerge, (2) we have a single
+             * thread for optimize, and the 'optimizeMutex' guarding this code, and (3) ConcurrentMergeScheduler
+             * syncs calls to findForcedMerges.
+             */
+            assert indexWriter.getConfig().getMergePolicy() instanceof ElasticsearchMergePolicy : "MergePolicy is " + indexWriter.getConfig().getMergePolicy().getClass().getName();
+            final ElasticsearchMergePolicy mp = (ElasticsearchMergePolicy) indexWriter.getConfig().getMergePolicy();
+            try {
                 if (upgrade) {
-                    logger.info("Starting upgrade of " + shardId);
-                    ((ElasticsearchMergePolicy) mp).setUpgradeInProgress(true);
+                    logger.info("starting segment upgrade");
+                    (mp).setUpgradeInProgress(true);
                 }
-
-                if (onlyExpungeDeletes) {
-                    indexWriter.forceMergeDeletes(false);
-                } else if (maxNumSegments <= 0) {
-                    indexWriter.maybeMerge();
-                } else {
-                    indexWriter.forceMerge(maxNumSegments, false);
+                store.incRef(); // increment the ref just to ensure nobody closes the store while we optimize
+                try {
+                    if (onlyExpungeDeletes) {
+                        assert upgrade == false;
+                        indexWriter.forceMergeDeletes(true);
+                    } else if (maxNumSegments <= 0) {
+                        assert upgrade == false;
+                        indexWriter.maybeMerge();
+                    } else {
+                        indexWriter.forceMerge(maxNumSegments, true);
+                    }
+                    if (flush) {
+                        flush(true, true, true);
```

Comment:
out of curiosity, why do we flush after the merge, instead of flushing before the merge to force writing new segments that can be considered for part of the forced merge?